src/osd/PrimaryLogPG.h

```diff
@@ -1622,18 +1622,25 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
       > reactions;
     struct ReservationCB : public Context {
       PrimaryLogPGRef pg;
+      OSDService *service;
       bool canceled;
-      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false) {}
+      ReservationCB(PrimaryLogPG *pg) : pg(pg), service(pg->osd),
+					canceled(false) {}
       void finish(int) override {
-	pg->lock();
+	PrimaryLogPG *pgp = pg.get();
+	if (service->is_stopping() || !pgp)
+	  return;
+	pgp->lock();
 	if (!canceled)
-	  pg->snap_trimmer_machine.process_event(SnapTrimReserved());
-	pg->unlock();
+	  pgp->snap_trimmer_machine.process_event(SnapTrimReserved());
+	pgp->unlock();
       }
       void cancel() {
 	assert(pg->is_locked());
 	assert(!canceled);
 	canceled = true;
+	pg->osd->snap_reserver.cancel_reservation(pg->get_pgid());
+	pg.reset();
```

Comment:
The whole point of this code is clearing out the pg reference so that it goes away before the Context is deleted. That's because we can't do a reliable deletion when we cancel a Context that's been given to an AsyncReserver as happens with this one .

That said, the race is pretty small and I don't think it's actually responsible for the bug report(s) I've been given so I'm trying to figure out other ref counting issues we may have.